x/concentrated-liquidity/internal/math/tick.go

```diff
@@ -1,52 +1,173 @@
+			exponentAtCurrentTick = exponentAtCurrentTick.Add(sdk.OneInt())
+			ticksPassed = ticksPassed.Add(geometricExponentIncrementDistanceInTicks.TruncateInt())
+		}
+	} else {
+		// We must decrement the exponentAtCurrentTick by one when traversing negative ticks in order to constantly step up in precision when going further down in ticks
+		// Otherwise, from tick 0 to tick -(geometricExponentIncrementDistanceInTicks), we would use the same exponent as the exponentAtPriceOne
+		exponentAtCurrentTick := exponentAtPriceOne.Sub(sdk.OneInt())
+		for totalPrice.GT(price) {
+			currentAdditiveIncrementInTicks = powTenBigDec(exponentAtCurrentTick)
+			maxPriceForcurrentAdditiveIncrementInTicks := osmomath.BigDecFromSDKDec(geometricExponentIncrementDistanceInTicks).Mul(currentAdditiveIncrementInTicks)
+			totalPrice = totalPrice.Sub(maxPriceForcurrentAdditiveIncrementInTicks.SDKDec())
+			exponentAtCurrentTick = exponentAtCurrentTick.Sub(sdk.OneInt())
+			ticksPassed = ticksPassed.Sub(geometricExponentIncrementDistanceInTicks.TruncateInt())
+		}
+	}
+	// Determine how many ticks we have passed in the exponentAtCurrentTick
+	ticksToBeFulfilledByExponentAtCurrentTick := osmomath.BigDecFromSDKDec(price.Sub(totalPrice)).Quo(currentAdditiveIncrementInTicks)
+
+	// Finally, add the ticks we have passed from the completed geometricExponent values, as well as the ticks we have passed in the current geometricExponent value
+	tickIndex := ticksPassed.Add(ticksToBeFulfilledByExponentAtCurrentTick.SDKDec().TruncateInt())
+
+	// Add a check to make sure that the tick index is within the allowed range
+	minTick, maxTick := GetMinAndMaxTicksFromExponentAtPriceOne(exponentAtPriceOne)
+	if tickIndex.LT(sdk.NewInt(minTick)) {
+		return sdk.Int{}, types.TickIndexMinimumError{MinTick: minTick}
+	}
+	if tickIndex.GT(sdk.NewInt(maxTick)) {
+		return sdk.Int{}, types.TickIndexMaximumError{MaxTick: maxTick}
+	}
+
+	return tickIndex, nil
+}
+
+// handleNegativeExponents treats negative exponents as 1/(10**|exponent|) instead of 10**-exponent
+// This is because the sdk.Dec.Power function does not support negative exponents
+func powTen(exponent sdk.Int) sdk.Dec {
+	if exponent.GTE(sdk.ZeroInt()) {
+		return sdkTenDec.Power(exponent.Uint64())
+	}
+	return sdk.OneDec().Quo(sdkTenDec.Power(exponent.Abs().Uint64()))
+}
+
+func powTenBigDec(exponent sdk.Int) osmomath.BigDec {
+	if exponent.GTE(sdk.ZeroInt()) {
+		return osmomath.NewBigDec(10).Power(osmomath.NewBigDec(exponent.Int64()))
+	}
+	return osmomath.OneDec().Quo(osmomath.NewBigDec(10).Power(osmomath.NewBigDec(exponent.Abs().Int64())))
+}
+
+// GetMinAndMaxTicksFromExponentAtPriceOne determines min and max ticks allowed for a given exponentAtPriceOne value
+// This allows for a min spot price of 0.000000000000000001 and a max spot price of 100000000000000000000000000000000000000 for every exponentAtPriceOne value
+func GetMinAndMaxTicksFromExponentAtPriceOne(exponentAtPriceOne sdk.Int) (minTick, maxTick int64) {
```

Comment:
Please add tests for this function

Reply:
https://github.com/osmosis-labs/osmosis/issues/3984